src/backend/distributed/transaction/remote_transaction.c

```diff
@@ -801,7 +810,15 @@ CoordinatedRemoteTransactionsPrepare(void)
 			continue;
 		}
 
-		FinishRemoteTransactionPrepare(connection);
+		/*
```

Comment:
Isn't the above `if (transaction->transactionState != REMOTE_TRANS_PREPARING)` already skips the read-only connections? If so, can we avoid the second call to `ConnectionModifiedPlacement ` as that might be an expensive call.

So, maybe convert it to an assert such as below?

```C
if (transaction->transactionState != REMOTE_TRANS_PREPARING)
{
    Assert (!ConnectionModifiedPlacement(connection))
    continue;
}
```

Also, in general `ConnectionModifiedPlacement` is pretty expensive check, iterating on all references. And this code-path is performance critical, so we'd want to avoid any additional overhead.

Is there a way to avoid the call with maybe (didn't check this idea in depth, just suggesting for further investigation and not to forget :) ):
```C
if (XactModificationLevel >= XACT_MODIFICATION_DATA && ConnectionModifiedPlacement(connection))
{
    ... StartRemoteTransactionPrepare...
}
```

Reply:
Depending on a previous statement seems a bit fragile to me, I understand the performance implication and the benefit of `if (XactModificationLevel >= XACT_MODIFICATION_DATA && ConnectionModifiedPlacement(connection))` but this seems to be complicating the code. Why don't we just have a flag in the connection struct that way we can access that field with 0 cost and use it whenever we need it without having to worry about any performance problems. That seems much simpler to me

Reply:
`XACT_MODIFICATION_DATA ` is not necessary and do not work anyway. So, ignore that please.


> Why don't we just have a flag in the connection struct that way we can access that field with 0 cost and use it whenever we need it without having to worry about any performance problems. 

I'm not completely against that, that can be an option. I see two aspects of it:

- Having an additional flag which keeps the same state as the others should not be required unless we have a really good reason. Because, it is always prone to bugs in the future (as someone might forget to update one of the flags).
- Do we really need this performance optimization? The most typical source of bugs is that we over engineer/optimize things that are not strictly necessary. **In any case,** we need some performance benchmarks for this PR. For example, do this PR add any performance penalty on HammerDB benchmarks? Overall, I'd prefer to give data-driven decisions here. And the benchmarks I'd do is the ones that require 2PC. Does HammerDB require? I think so. 


So, lets make sure that we run at least HammerDB and pgbench's tpcb-like benchmarks with `ConnectionModifiedPlacement ()` calls in the code vs master.  And, similarly some more tests where we  iterate inside the `ConnectionModifiedPlacement ()`, such as the following on a single thread pgbench:
```
-- set the pool size to 1 and make sure that all
-- placements are accessed over the same conection
SET citus.adaptive_executor_pool_size TO 1;
DELETE FROM table;
```

With this data, we can have a better idea on whether we need this optimizations or not

Reply:
tpcc results:
https://github.com/citusdata/release-test-results/tree/citus_github_push/citusbot_tpcc_benchmark_rg/02_03_2021_1612347049/02_03_2021_1612361561/hammerdb/02_03_2021_1612361561

399K current_branch vs 392K master
seems like no regression.